ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Handlers.hs

```diff
@@ -144,3 +152,49 @@ handlerRequestNext currentIntersection blockTree points =
     TipPoint tip' = tip points
 
     trace = tell . pure
+
+handlerBlockFetch ::
+  forall m .
+  IOLike m =>
+  BlockTree TestBlock ->
+  ChainRange (Point TestBlock) ->
+  AdvertisedPoints ->
+  STM m (Maybe BlockFetch, [String])
+handlerBlockFetch blockTree (ChainRange from to) AdvertisedPoints {header = HeaderPoint hp, block = BlockPoint bp} =
+  runWriterT (serveFromBpFragment (AF.sliceRange bpChain from to))
+  where
+    -- Check whether the requested range is contained in the fragment before the block point.
+    -- We may only serve the full range; if the server has only some of the blocks, it must refuse.
+    serveFromBpFragment = \case
+      Just slice -> do
+        trace ("Sending slice " ++ terseFrag slice)
+        pure (Just (StartBatch (AF.toOldestFirst slice)))
+      Nothing    -> do
+        trace ("Waiting for next tick for range: " ++ tersePoint from ++ " -> " ++ tersePoint to)
+        -- If we cannot serve blocks from the BP chain, decide whether to yield control to the scheduler
+        -- or serve blocks anyway.
+        --
+        -- If the @to@ point is not part of the HP chain but BP is, we must have switched to a fork without ensuring
+        -- that BP advances to the last HP advertised on the old chain.
+        -- This should be a precondition violation, but because it would make the schedule generator more complex, we
+        -- simply serve the missing blocks anyway here.
+        --
+        -- Otherwise, we simply have to wait for BP to advance sufficiently, and we block without sending
+        -- a message, to simulate a slow response.
+        case not (AF.withinFragmentBounds to hpChain) && AF.withinFragmentBounds (toPoint bp) hpChain of
+          True ->
+            case AF.sliceRange (fragmentUpTo "requested point" to) from to of
+              Just slice -> pure (Just (StartBatch (AF.toOldestFirst slice)))
+              Nothing    -> error "BlockFetch: Client requested "
```

Comment:
Here I'd expect:

if the request is not in the chain of the HP serve the blocks (or send NoBlock)
else wait for the BP to be updated (this is the meaning of yielding Nothing IIUC)

I don't understand the relevance of checking if the BP is in the chain of HP.

Reply:
oh I thought that was what you suggested – if BP is on HP's chain, but `to` isn't, we definitely have switched both to a fork without letting BP catch up.

Unless BP is so far back that it's before the fork intersection…what do you think?

Reply:
FWIW if we want to precisely detect it we could additionally check whether BP is on the chain that leads up to `to`